lib/protobuf/field/message_field.rb

```diff
@@ -35,8 +35,8 @@ def coerce!(value)
           nil
         elsif value.is_a?(type_class)
           value
-        elsif value.respond_to?(:to_proto)
-          value.to_proto
+        elsif value.respond_to?(:to_proto) && (proto_value = value.to_proto).is_a?(type_class)
```

Comment:
Is this check necessary? If so, the complexity of a compound conditional that also sets a variable that is used as the return value is something we try to eliminate. If the goal is to avoid having to write the failure line twice, then I'd rewrite the method:

```ruby
def coerce!(value)
  return nil if value.nil?

  coerced_value = if value.respond_to?(:to_proto)
                    value.to_proto
                  elsif value.respond_to?(:to_hash)
                    type_class.new(value.to_hash)
                  else
                    value
                  end
  
  return coerced_value if coerced_value.is_a?(type_class)

  fail TypeError, "Expected value of type '#{type_class}' for field #{name}, but got '#{value.class}'"
end
```

To me, this is much clearer. It basically has three steps:

1. Ignore nil values
2. Try and coerce the value
3. Fail unless the coerced value is the correct type

What do you think?